### PR TITLE
Add derived_attribute capabilities

### PR DESCRIPTION
**Pull request by andypinz**
Add a new type for erlen - derived_attribute, that allows the user to derive a value from other schema values.

### CODE REVIEW

File: lib/erlen/schema/base.rb [hunk 1]
```diff
@@ -316,6 +337,10 @@ def __has_attribute(name)
       !__find_attribute_name(name).nil?
     end
 
+    def __has_derived_attribute(name)
```
Comment (iande) [inline lib/erlen/schema/base.rb]: Add `?` to the end of this method name, since it is a predicate method.

File: lib/erlen/schema/base.rb [hunk 2]
```diff
@@ -378,5 +403,11 @@ def __find_attribute_value_by_name(name)
       val.is_a?(Undefined) ? nil : val
     end
 
+    def __find_derived_attribute_value_by_name(name)
+      val = self.class.schema_derived_attributes[name].derive_value(self)
+
+      # We don't want to expose Undefined to the outside world, nil it the logical equal
```
Comment (iande) [inline lib/erlen/schema/base.rb]: "... nil it the logical ..." => "... nil is the logical ..." ?

File: lib/erlen/schema/base.rb [hunk 3]
```diff
@@ -378,5 +403,11 @@ def __find_attribute_value_by_name(name)
...
+      val.is_a?(Undefined) ? nil : val
```
Comment (iande) [inline lib/erlen/schema/base.rb]: Is it possible for `val` to be an instance of `Undefined`?

File: spec/erlen/schema/base_spec.rb [hunk 4]
```diff
@@ -268,8 +291,14 @@ class TestContextSchema < TestBaseSchema
       payload = TestBaseSchema.import(TestObj.new)
       data = payload.to_data
 
-      expect(data['foo']).to eq('bar')
-      expect(data['custom']).to eq(nil)
+      expect(data['derived_attr']).to eq('bar10')
```
Comment (iande) [inline spec/erlen/schema/base_spec.rb]: Why remove the other expectations that were here?

Comment (iande) [reply spec/erlen/schema/base_spec.rb]: If it's due to them being tested in the block above, I'll add some context: `to_hash` was the original method for converting a schema to a simple structure suitable for conversion to JSON; however, this name proved to be problematic for `ArrayOf` types, so `to_data` was introduced and `to_hash` was deprecated.

File: spec/erlen/schema/base_spec.rb [hunk 5]
```diff
@@ -268,8 +291,14 @@ class TestContextSchema < TestBaseSchema
...
+    end
+
+    it 'includes derived attribute' do
+      payload = TestBaseSchema.new
+      data = payload.to_data
+      expect(data.include?('foo')).to be_falsey
+      expect(data.include?('custom')).to be_falsey
```
Comment (iande) [inline spec/erlen/schema/base_spec.rb]: This doesn't appear to be verifying that it "includes derived attribute".

File: lib/erlen/schema/base.rb [hunk 6]
```diff
@@ -63,6 +66,17 @@ def validate(message, &blk)
         validator_procs << [message, blk]
       end
 
+      # Defines a derived attribute for the schema. Must specify name. Must
+      # specify code block to derive the value
+      #
+      # @param name [String, Symbol] a simple name for the derived attribute
+      # @param blk [Proc] the code block that derives the value from the
```
Comment (iande) [inline lib/erlen/schema/base.rb]: Should include the `type` param in the YARD comments as well as per the [attribute method](https://github.com/Hireology/erlen/blob/415917ed4476515db13eb2765a5dbca26dbec963/lib/erlen/schema/base.rb#L37)

File: spec/erlen/schema/derived_attribute_spec.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
+require 'spec_helper'
+
+describe Erlen::Schema::DerivedAttribute do
+  subject { described_class }
+
+  describe '#initialize' do
+    it 'sets all the values' do
+      schema = DerivedSchema.new({})
+      attr = subject.new(:full_name, Type, proc do |s|
```
Comment (iande) [inline spec/erlen/schema/derived_attribute_spec.rb]: I think `Type` here should be replaced with `String`.

File: lib/erlen/schema/base.rb (hunk 3 above)
Comment (andypinz) [reply lib/erlen/schema/base.rb]: Good point. It isn't, I think I can safely remove this line.

File: spec/erlen/schema/base_spec.rb (hunk 5 above)
Comment (andypinz) [reply spec/erlen/schema/base_spec.rb]: You are very correct. Not sure how I missed that one.

File: spec/erlen/schema/base_spec.rb (hunk 4 above)
Comment (andypinz) [reply spec/erlen/schema/base_spec.rb]: This is a case of me copypasting and changing the wrong test description. We can zoom about this later (I will fix the incorrect test description).

File: lib/erlen/schema/base.rb (hunk 6 above)
Comment (andypinz) [reply lib/erlen/schema/base.rb]: 👍

File: spec/erlen/schema/derived_attribute_spec.rb (hunk 7 above)
Comment (andypinz) [reply spec/erlen/schema/derived_attribute_spec.rb]: 👍

**Pull request merged**